servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -304,33 +304,32 @@ private boolean tryDecrementPendingDemand() {
         }
 
         private void tryEmitItem(Object item, FlatMapPublisherSubscriber<T, R> subscriber) {
-            if (tryAcquireLock(emittingLockUpdater, this)) { // fast path. no concurrency, try to skip the queue.
-                boolean mappedSourcesCompleted = false;
+            // We can skip the queue if the following conditions are meet:
+            // 1. There is downstream requestN demand.
+            // 2. The mapped subscriber doesn't have any signals already in the queue. We only need to preserve the
+            //    ordering for each mapped source, and there is no "overall" ordering.
+            // 3. We don't concurrently invoke the downstream subscriber. Concurrency control is provided by the
+            //    emitting lock.
+            final boolean needsDemand;
+            if (subscriber.hasSignalsQueued() || ((needsDemand = needsDemand(item)) && !tryDecrementPendingDemand())) {
+                subscriber.markSignalsQueued();
+                enqueueAndDrain(item);
+            } else if (item == MAPPED_SOURCE_COMPLETE) {
+                requestMoreFromUpstream(1);
+            } else if (tryAcquireLock(emittingLockUpdater, this)) { // fast path. no concurrency, try to skip the queue.
                 try {
-                    // We can skip the queue if the following conditions are meet:
-                    // 1. There is downstream requestN demand.
-                    // 2. The mapped subscriber doesn't have any signals already in the queue. We only need to preserve
-                    //    the ordering for each mapped source, and there is no "overall" ordering.
-                    // 3. We don't concurrently invoke the downstream subscriber. Concurrency control is provided by the
-                    //    emitting lock.
-                    if (subscriber.hasSignalsQueued() || (needsDemand(item) && !tryDecrementPendingDemand())) {
-                        subscriber.markSignalsQueued();
-                        enqueueItem(item);
```

Comment:
the issue was here if the item doesn't require demand we should force a drain as we can't rely upon more demand to trigger a domain. however I refactored this method a bit to reduce the critical area and make the drain less conditional (e.g. now we can just use `enqueueAndDrain`).